packages/@react-aria/dnd/src/DropTargetKeyboardNavigation.ts

```diff
@@ -0,0 +1,273 @@
+    while (lastKey != null) {
+      prevKey = lastKey;
+      let node = collection.getItem(lastKey);
+      lastKey = node?.parentKey;
+    }
+
+    if (prevKey != null) {
+      return {
+        type: 'item',
+        key: prevKey,
+        dropPosition: 'after'
+      };
+    }
+
+    return null;
+  }
+
+  if (target.type === 'item') {
+    let prevKey: Key | null | undefined = null;
+    if (horizontal) {
+      prevKey = horizontal === 'left' ? keyboardDelegate.getKeyLeftOf?.(target.key) : keyboardDelegate.getKeyRightOf?.(target.key);
+    } else {
+      prevKey = keyboardDelegate.getKeyAbove?.(target.key);
+    }
+    let prevCollectionKey = collection.getKeyBefore(target.key);
+
+    // If the keyboard delegate did not move to the next key in the collection,
+    // jump to that key with the same drop position. Otherwise, try the other
+    // drop positions on the current key first.
+    if (prevKey != null && prevKey !== prevCollectionKey) {
+      return {
+        type: 'item',
+        key: prevKey,
+        dropPosition: target.dropPosition
+      };
+    }
+
+    switch (target.dropPosition) {
+      case 'before': {
+        // Move after the last child of the previous item.
+        let targetNode = collection.getItem(target.key);
+        if (targetNode && targetNode.prevKey != null) {
+          let lastChild = getLastChild(collection, targetNode.prevKey);
+          if (lastChild) {
+            return lastChild;
+          }
+        }
+
+        if (prevKey != null) {
+          return {
+            type: 'item',
+            key: prevKey,
+            dropPosition: 'on'
+          };
+        }
+
+        return {
+          type: 'root'
+        };
+      }
+      case 'on': {
+        return {
+          type: 'item',
+          key: target.key,
+          dropPosition: 'before'
+        };
+      }
+      case 'after': {
+        // Move after the last child of this item.
+        let lastChild = getLastChild(collection, target.key);
+        if (lastChild) {
+          return lastChild;
+        }
+
+        return {
+          type: 'item',
+          key: target.key,
+          dropPosition: 'on'
+        };
+      }
+    }
+  }
+
+  if (target.type !== 'root') {
+    return {
+      type: 'root'
+    };
+  }
+
+  return null;
+}
+
+function getLastChild(collection: Collection<Node<unknown>>, key: Key): DropTarget | null {
+  // getChildNodes still returns child tree items even when the item is collapsed.
+  // Checking if the next item has a greater level is a silly way to determine if the item is expanded.
+  let targetNode = collection.getItem(key);
+  let nextKey = collection.getKeyAfter(key);
+  let nextNode = nextKey != null ? collection.getItem(nextKey) : null;
+  if (targetNode && nextNode && nextNode.level > targetNode.level) {
+    let children = getChildNodes(targetNode, collection);
+    let lastChild: Node<unknown> | null = null;
+    for (let child of children) {
```

Comment:
nit pick optimisation, loop by index from the last item back to the first, should on average hit the last child faster and may not need to traverse the whole thing because you could early return

Reply:
I can include this in a follow-up

Reply:
`children` is an Iterable, not an array. So there is no way to iterate backward, without first converting to an Array (which itself requires iterating forward).

For new collections we could use `lastChildKey` and iterate backward through the `prevKey`s, but this is not available in old collections.